internal/training/grpo_engine/kernels.c

```
@@ -0,0 +1,764 @@
+    float result;
+    memcpy(&result, &f, sizeof(float));
+    return result;
+}
+
+/* ─── Q4_K Dequantization ─── */
+/* Real Q4_K super-block (144 bytes for 256 elements):
+ *   half d (2B) + half dmin (2B) + uint8 scales[12] + uint8 qs[128]
+ * The 128 bytes of qs hold 256 4-bit values (2 per byte).
+ * The 12 bytes of scales encode per-sub-block (16 sub-blocks of 16 elements) scale factors.
+ * Simplified: use global d/dmin without sub-block scales (lossy but functional). */
+static inline float dequant_q4k_element(const uint8_t *block, int idx) {
```

Cppcheck (2.21.0)

[style] The function 'dequant_q4k_element' is never used.

(unusedFunction)